validator/inputs/writer.go

```diff
@@ -0,0 +1,129 @@
+package inputs
+
+import (
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/offchainlabs/nitro/validator/server_api"
+)
+
+// Writer is a configurable writer of InputJSON files.
+//
+// The default Writer will write to a path like:
+//
+//	$HOME/.arbuitrum/validation-inputs/<YYYMMDD_HHMMSS>/block_inputs_<id>.json
+//
+// The path can be nested under a slug directory so callers can provide a
+// recognizable name to differentiate various contexts in which the InputJSON
+// is being written. If the Writer is configured by calling SetSlug, then the
+// path will be like:
+//
+//	$HOME/.arbuitrum/validation-inputs/<slug>/<YYYMMDD_HHMMSS>/block_inputs_<id>.json
+//
+// The inclusion of a timestamp directory is on by default to avoid conflicts which
+// would result in files being overwritten. However, the Writer can be configured
+// to not use a timestamp directory.  If the Writer is configured by calling
+// SetUseTimestampDir(false), then the path will be like:
+//
+//	$HOME/.arbuitrum/validation-inputs/<slug>/block_inputs_<id>.json
+//
+// Finally, to give complete control to the clients, the base directory can be
+// set directly with SetBaseDir. In which case, the path will be like:
+//
+//	<baseDir>/block_inputs_<id>.json
+//	  or
+//	<baseDir>/<slug>/block_inputs_<id>.json
+//	  or
+//	<baseDir>/<slug>/<YYYMMDD_HHMMSS>/block_inputs_<id>.json
+type Writer struct {
+	clock           Clock
+	baseDir         string
+	slug            string
+	useTimestampDir bool
+}
+
+// Clock is an interface for getting the current time.
+type Clock interface {
+	Now() time.Time
+}
+
+type realClock struct{}
+
+func (realClock) Now() time.Time {
+	return time.Now()
+}
+
+// NewWriter creates a new Writer with default settings.
+func NewWriter() (*Writer, error) {
+	homeDir, err := os.UserHomeDir()
+	if err != nil {
+		return nil, err
+	}
+	baseDir := fmt.Sprintf("%s/.arbitrum/validation-inputs", homeDir)
+	return &Writer{
+		clock:           realClock{},
+		baseDir:         baseDir,
+		slug:            "",
+		useTimestampDir: true}, nil
+}
+
+// SetClockForTesting sets the clock used by the Writer.
+//
+// This is only intended for testing.
+func (w *Writer) SetClockForTesting(clock Clock) *Writer {
```

Comment:
this function is only used by tests in the same module so doesn't need to be public